File: solutions/python/say/1/say.py
def say(number):
    digits_words = ['one','two','three','four','five','six','seven','eight','nine','ten','eleven','twelve','thirteen','fourteen','fifteen','sixteen','seventeen','eighteen','nineteen']
    tens_words = ['','twenty','thirty','forty','fifty','sixty','seventy','eighty','ninety']
    columns_words = ['','thousand','million','billion']

    if number == 0: return 'zero'
    if 0 > number or number > 999999999999: raise ValueError('input out of range')

    number_reversed = str(number)[::-1]
    number_split = [str(number_reversed)[i:i+3][::-1] for i in range(0, len(str(number_reversed)), 3)]
    
    def three_digit_fragment(number):
        output = ''
        last_2_digits = int(str(number)[-2:])
    
        if 0 < last_2_digits < 20:
            output = digits_words[last_2_digits - 1]
        elif last_2_digits > 19:
            output = tens_words[int(str(number)[-2:-1])-1]
            if int(str(number)[-1]) != 0: output = output + '-' + digits_words[int(str(number)[-1:])-1]
        if number > 99:
            output = digits_words[int(str(number)[-3:-2])-1] + ' hundred ' + output
            
        return output.strip()

    output = ''

    for fragment in number_split:
        output = three_digit_fragment(int(fragment)) + ' ' + columns_words[number_split.index(fragment)] + ' ' + output

    return output.strip()

File: solutions/cpp/say/1/say.cpp
#include "say.h"

namespace say {
    std::string in_english(long long unsigned int number_) {
        if (number_ >= 1000000000000) {
            throw std::domain_error("number out of bounds");
        }
        if (number_ == 0) return "zero";

        long long unsigned int number = number_;
        std::string output{};

        // billions
        if (number > 999999999) {
            int chunk_billions = (number % 1000000000000) / 1000000000;
            output += say::convert_chunk(chunk_billions) + " billion";
            if (number % 1000000000 != 0) output += " ";
        }
        
        // millions
        if (number % 10000000 > 999999) {
            int chunk_millions = (number % 1000000000) / 1000000;
            output += say::convert_chunk(chunk_millions) + " million";
            if (number % 1000000 != 0) output += " ";
        }
        
        // thousands
        if (number % 10000 > 999) {
            int chunk_thousands = (number % 1000000) / 1000;
            output += say::convert_chunk(chunk_thousands) + " thousand";
            if (number % 1000 != 0) output += " ";
        }

        // < 1000
        output += say::convert_chunk(number % 1000);
        
        return output;
    }

        std::string convert_chunk(int number) {
            
            std::vector<std::string> under_20 = {
            "", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", 
            "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"};
            std::vector<std::string> tens = {"", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"};
            
            std::string chunk_string{};
            // hundreds
            if (number % 1000 > 99) {
                chunk_string += under_20[(number % 1000) / 100] + " hundred";
                if (number % 100 != 0) chunk_string += " ";
            }
            
            // right 2 digits
            if ((number % 100) < 20) {
                chunk_string += under_20[number % 100];
            } else {
                if (19 < (number % 100) and (number % 100) < 100) chunk_string += tens[(number % 100) / 10];
                number = number % 10; // need this truncation or errors
                if (number % 10 != 0) chunk_string += "-" + under_20[number];
            }

            return chunk_string;
        }

}  // namespace say
